### PR TITLE
Migrate to IPFabric Client Library

### PR DESCRIPTION
**Pull request by h4ndzdatm0ld**
A start to migrate the API client the IPFabric maintained python library.

TODO:

- [x] - Linting, the usual
- [x] - Replace the `get_vlans` call with whatever the equivalent is in the new library
- [x] - Look into maybe adding more valid unit tests with more data
- [x] - Add a unit test to validate the logic of serial number. Even though char limit will be raised in later Nautobot version, we want it backwards compatible

### CODE REVIEW

File: nautobot_ssot_ipfabric/jobs.py [hunk 1]
```diff
@@ -131,8 +130,7 @@ def log_debug(self, message):
 
     def sync_data(self):
         """Sync a device data from IP Fabric into Nautobot."""
-        client = IpFabricClient(IPFABRIC_HOST, IPFABRIC_API_TOKEN)
-
+        client = IPFClient(IPFABRIC_HOST, token=IPFABRIC_API_TOKEN, verify=False, timeout=15)
```
Comment (h4ndzdatm0ld) [inline nautobot_ssot_ipfabric/jobs.py:133]: verify/timeout should prob come from settings

File: nautobot_ssot_ipfabric/diffsync/adapter_ipfabric.py [hunk 2]
```diff
@@ -76,9 +80,12 @@ def load_device_interfaces(self, device_model, interfaces, device_primary_ip):
     def load(self):
         """Load data from IP Fabric."""
         self.load_sites()
-        devices = self.client.get_device_inventory()
-        interfaces = self.client.get_interface_inventory()
+        devices = self.client.inventory.devices.all()
+        interfaces = self.client.inventory.interfaces.all()
+
+        # TODO: HERE, find vlan getter
         vlans = self.client.get_vlans()
```
Comment (h4ndzdatm0ld) [inline nautobot_ssot_ipfabric/diffsync/adapter_ipfabric.py:87]: Find this call

**Comment by pke11y** [comment]
Looks straight forward. 
@justinjeffery-ipf can you take a look?

**Review by pke11y — APPROVED** [review]

**Comment by justinjeffery-ipf** [comment]
An exception occurred: DataError: value too long for type character varying(50)

Action: DiffSyncActions.UPDATE
Synced_object: vlan 1ry firewall__35HEADOFFICE


`{
  "+": {
    "vid": 5,
    "description": "VLAN ID: 5"
  },
  "-": {
    "vid": 4,
    "description": "VLAN ID: 4"
  }
}`


`Traceback (most recent call last):
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/models/query.py", line 573, in get_or_create
    return self.get(**kwargs), False
  File "/opt/nautobot/lib/python3.9/site-packages/cacheops/query.py", line 353, in get
    return qs._no_monkey.get(qs, *args, **kwargs)
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/models/query.py", line 429, in get
    raise self.model.DoesNotExist(
nautobot.dcim.models.devices.Device.DoesNotExist: Device matching query does not exist.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.StringDataRightTruncation: value too long for type character varying(50)


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/nautobot/lib/python3.9/site-packages/nautobot_ssot/jobs/base.py", line 326, in run
    self.sync_data()
  File "/opt/nautobot/lib/python3.9/site-packages/nautobot_ssot_ipfabric/jobs.py", line 185, in sync_data
    dest.sync_from(ipfabric_source)
  File "/opt/nautobot/lib/python3.9/site-packages/diffsync/__init__.py", line 525, in sync_from
    result = syncer.perform_sync()
  File "/opt/nautobot/lib/python3.9/site-packages/diffsync/helpers.py", line 326, in perform_sync
    changed |= self.sync_diff_element(element)
  File "/opt/nautobot/lib/python3.9/site-packages/diffsync/helpers.py", line 390, in sync_diff_element
    changed |= self.sync_diff_element(child, parent_model=dst_model)
  File "/opt/nautobot/lib/python3.9/site-packages/diffsync/helpers.py", line 366, in sync_diff_element
    changed, modified_model = self.sync_model(src_model=src_model, dst_model=dst_model, ids=ids, attrs=attrs)
  File "/opt/nautobot/lib/python3.9/site-packages/diffsync/helpers.py", line 415, in sync_model
    dst_model = self.model_class.create(diffsync=self.dst_diffsync, ids=ids, attrs=attrs)
  File "/opt/nautobot/lib/python3.9/site-packages/nautobot_ssot_ipfabric/diffsync/diffsync_models.py", line 191, in create
    new_device, _ = NautobotDevice.objects.get_or_create(
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/models/query.py", line 576, in get_or_create
    return self._create_object_from_params(kwargs, params)
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/models/query.py", line 610, in _create_object_from_params
    obj = self.create(**params)
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/models/query.py", line 447, in create
    obj.save(force_insert=True, using=self.db)
  File "/opt/nautobot/lib/python3.9/site-packages/nautobot/dcim/models/devices.py", line 735, in save
    super().save(*args, **kwargs)
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/models/base.py", line 753, in save
    self.save_base(using=using, force_insert=force_insert,
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/models/base.py", line 790, in save_base
    updated = self._save_table(
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/models/base.py", line 895, in _save_table
    results = self._do_insert(cls._base_manager, using, fields, returning_fields, raw)
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/models/base.py", line 933, in _do_insert
    return manager._insert(
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/models/query.py", line 1254, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1397, in execute_sql
    cursor.execute(sql, params)
  File "/opt/nautobot/lib/python3.9/site-packages/cacheops/transaction.py", line 93, in execute
    result = self._no_monkey.execute(self, sql, params)
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/opt/nautobot/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.DataError: value too long for type character varying(50)
`

**Comment by justinjeffery-ipf** [comment]
Logs are not working either.  Saying a failure but tells me no failure.

image

**Comment by chadell** [comment]
thanks @justinjeffery-ipf for the feedback.

**Comment by pke11y** [comment]
I ran this against a clean Nautobot and it worked.
@justinjeffery-ipf I'm guessing you had data in your Nautobot before syncing from IP Fabric?

**Comment by justinjeffery-ipf** [comment]
Some yes but not everything as Azure serial numbers are too long to import.  I deleted that one VLAN entry but I don't think that was the issue as it is failing for some unknown reason.

**Pull request merged**